src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -780,7 +781,7 @@ private void CompletionCallback(object waitHandleContext, bool wasSignaled)
             {
                 // Check the exited event that we get from the threadpool
                 // matches the event we are waiting for.
-                if (waitHandleContext != _waitHandle)
+                if (waitHandleContext != _completionCallbackContext)
```

Comment:
Could you just put this check behind a property implemented differently for Windows/Unix, to avoid needing this field on Windows, too?